### PR TITLE
[ENH] BIDS software development guidelines appendix

### PR DESCRIPTION
**Pull request by yarikoptic**
Related (possibly to be reflected in that doc):
- https://github.com/bids-standard/bids-specification/issues/56

This PR is primarily to palpate interest and possible objections

Additional TODOs known needed to be done
- [ ] verify that linkage to those other documents works (manually until #79 will come back from the ashes)

### CODE REVIEW

**Comment by sappelhoff** [comment]
Looks like a good thing to have. It reminds me a bit of the [BEP LEAD GUIDELINES](https://docs.google.com/document/d/1pWmEEY-1-WuwBPNy5tDAxVJYQ9Een4hZJM06tQZg8X4/edit) (formerly Contributor Guide).

Both documents seem to be on the same logical level, so I would propose to either
- make the software development guide a GoogleDoc
- import the BEP LEAD GUIDELINES into github as an appendix

File: src/99-appendices/09-software-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+BIDS is an evolving standard, and although we strive to clearly and unambiguously
+describe all entities and filenames convention, retain backward compatibility, 
+and version every release of the specification, some aspects could remain not
+clarified and open for interpretation or future change.  E.g., the set of 
+allowed characters in `<label>` might eventually change [ISSUE-226] or 
+[inheritance principle](../02-common-principles.md#the-inheritance-principle)
+could be further formalized (see [ISSUE-102]).
+It is RECOMMENDED that BIDS supporting applications are coded adhering 
+to [Postel's principle](https://en.wikipedia.org/wiki/Robustness_principle)
+which is driving TCP implementations to *"be conservative in what you do, be
```
Comment (nicholst) [inline src/99-appendices/09-software-guidelines.md]: For the sake of clarity and minimising the nerd factor, I'd remove "driving TCP implementations"

File: src/99-appendices/09-software-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+allowed characters in `<label>` might eventually change [ISSUE-226] or 
+[inheritance principle](../02-common-principles.md#the-inheritance-principle)
+could be further formalized (see [ISSUE-102]).
+It is RECOMMENDED that BIDS supporting applications are coded adhering 
+to [Postel's principle](https://en.wikipedia.org/wiki/Robustness_principle)
+which is driving TCP implementations to *"be conservative in what you do, be 
+liberal in what you accept from others."*
+
+
+[ISSUE-226]: https://github.com/bids-standard/bids-specification/issues/226
```
Comment (nicholst) [inline src/99-appendices/09-software-guidelines.md]: I don't understand the role of these links... we reference issues inside the spec?  Seems fragile.

**Comment by sappelhoff** [comment]
@franklin-feingold @effigies do yo want to chime in on this PR?

Do you think the contents should be added to the spec? ... or in a different/separate Google Doc?

see also my comment above. @nicholst your comments were more of a concrete nature - but perhaps you also want to voice your opinion on where (and if at all) to include this in general.

**Comment by nicholst** [comment]
My comments all seem very petty now :)

Anyway, I *do* see the role and value of guidelines for BIDS *software*, and it makes sense that it is distinct from  BEP LEAD GUIDELINES, right?

E.g., if I'm someone wanting to write software for BIDS 1.0, what does that have to do with a BEP?

So... unless there's another concrete suggestion, here as an appendix seems as good as any place.

**Comment by sappelhoff** [comment]
> Anyway, I do see the role and value of guidelines for BIDS software, and it makes sense that it is distinct from BEP LEAD GUIDELINES, right?

I agree that these should be distinct documents - but I see where the confusion is coming from.

I was simply saying that if we make THESE guidelines (for software) a part of the spec ... then why shouldn't the BEP-Lead-guidelines be a part of the spec as well?

This is just an argument for consistency, and I currently tend towards preferring putting the software guidelines in a google doc (until they grow more substantially) - but I am happy to hear other opinions.

**Comment by franklin-feingold** [comment]
My two cents on the point raised in this discussion:

Regarding having this part of the appendix - In the current makeup of the appendix, this document may be better suited to be part of [the specification repository](https://github.com/bids-standard/bids-specification) (temporarily) rather than in the specification RTD. This would be great in relation to information/repository for contributors (similar to the starter kit, but for contributors. perhaps establishing a need for a contributors repo?). There is infrastructure to be built out to support this type of information effectively (and even overall better establishing the process, documenting, and simplifying the pathway from user to contributor). In the meantime, perhaps this can be part of this repository. It can be in markdown, but if there is dynamic action that arises it can be ported over to google docs and the more rapid commenting/development can happen more seamlessly there. Google doc vs markdown I don't feel strongly either way.

In general, having a set of guidelines to uniformly build out software associated with BIDS will be really important as we move into wrapping up Derivatives, Statistical and Computational Models.

File: src/99-appendices/09-software-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+BIDS is an evolving standard, and although we strive to clearly and unambiguously
+describe all entities and filenames convention, retain backward compatibility, 
+and version every release of the specification, some aspects could remain not
+clarified and open for interpretation or future change.  E.g., the set of 
+allowed characters in `<label>` might eventually change [ISSUE-226] or 
+[inheritance principle](../02-common-principles.md#the-inheritance-principle)
+could be further formalized (see [ISSUE-102]).
+It is RECOMMENDED that BIDS supporting applications are coded adhering 
+to [Postel's principle](https://en.wikipedia.org/wiki/Robustness_principle)
```
Comment (jasmainak) [inline src/99-appendices/09-software-guidelines.md]: It's not clear what this means in the BIDS context

**Comment by sappelhoff** [comment]
I went ahead and copied the contents to a google doc [BIDS Software Development Guidelines](https://docs.google.com/document/d/1_A1GUHAbczHL6WnEw3b8_8_4PXb1A4wZ4Ckrjgapt5c/edit?usp=sharing)

this document may be more easily shared and extended by other contributors.

If it grows and we finally decide what to do with it, we can reopen this PR.

Feel free to challenge this decision.

**Comment by yarikoptic** [comment]
- "Daddy, daddy!? what is a Google Doc?"
- "Google Doc is where all good but unfinished ideas are moved to let them RiP" ;)

**Pull request closed without merging**